docs/rfcs/040-Endpoint-Persistent-Unlogged-Files-Storage.md

```diff
@@ -0,0 +1,288 @@
+# Memo: Endpoint Persistent Unlogged Files Storage
+Created on 2024-11-05
+Implemented on N/A
+
+## Summary
+A design for a storage system that allows storage of files required to make
+Neon's Endpoints have a better experience at or after a reboot.
+
+## Motivation
+Several systems inside PostgreSQL (and Neon) need some persistent storage for
+optimal workings across reboots and restarts, but still work without.
+Examples are the cumulative statistics file in `pg_stat/global.stat`,
+`pg_stat_statements`' `pg_stat/pg_stat_statements.stat`, and `pg_prewarm`'s
+`autoprewarm.blocks`.  We need a storage system that can store and manage
+these files for each Endpoint, without granting users access to an unlimited
+storage device.
+
+## Goals
+- Store known files for Endpoints with reasonable persistence.  
+  _Data loss in this service, while annoying and bad for UX, won't lose any
+  customer's data._
+
+## Non Goals (if relevant)
+- This storage system does not need branching, file versioning, or other such
+  features. The files are as ephemeral to the timeline of the data as the
+  Endpoints that host the data.
+- This storage system does not need to store _all_ user files, only 'known'
+  user files.
+- This storage system does not need to be hosted fully inside Computes.  
+  _Instead, this will be a separate component similar to Pageserver,
+  SafeKeeper, the S3 proxy used for dynamically loaded extensions, etc._
+
+## Impacted components (e.g. pageserver, safekeeper, console, etc)
+- Compute needs new code to load and store these files in its lifetime.
+- Console or Control Plane needs to consider this new storage system when
+  signalling the deletion of an Endpoint, Timeline, or Tenant.
+
+A new service is created: the Endpoint Persistent Unlogged Files Storage
+service.  This could be integrated in e.g. Pageserver or Control Plane, or a
+separately hosted service.
+
+## Proposed implementation
+Endpoint-related data files are managed by a newly designed service (which
+optionally is integrated in an existing service like Pageserver or Control
+Plane), which stores data directly into S3 or any blob storage of choice.
+
+Upon deletion of the Endpoint, or reassignment of the endpoint to a different
+branch, this ephemeral data is dropped: the data stored may not match the
+state of the branch's data after reassignment, and on endpoint deletion the
+data won't have any use to the user.
+
+Compute gets credentials which it can use to authenticate to this new service
+and retrieve and store data associated with that endpoint.
```

Comment:
I think we could elaborate on that, i.e. that we will use JWTs with tenant+timeline IDs, which both provides good tenants isolation and adds an additional protection layer for different timelines to do not mess with each other data